hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/Dispatcher.java

```diff
@@ -60,137 +64,177 @@ public Dispatcher(
-            case TOKEN_UNPAUSE -> handlers.tokenUnpauseHandler().preHandle(transactionBody, payer);
-
-            case UTIL_PRNG -> handlers.utilPrngHandler().preHandle(transactionBody, payer);
-
-            case SYSTEMDELETE -> switch (transactionBody.getSystemDelete().getIdCase()) {
-                case CONTRACTID -> handlers.contractSystemDeleteHandler()
-                        .preHandle(transactionBody, payer);
-                case FILEID -> handlers.fileSystemDeleteHandler().preHandle(transactionBody, payer);
-                case ID_NOT_SET -> throw new IllegalArgumentException(
-                        "SystemDelete without IdCase");
-            };
-            case SYSTEMUNDELETE -> switch (transactionBody.getSystemUndelete().getIdCase()) {
-                case CONTRACTID -> handlers.contractSystemUndeleteHandler()
-                        .preHandle(transactionBody, payer);
-                case FILEID -> handlers.fileSystemUndeleteHandler()
-                        .preHandle(transactionBody, payer);
-                case ID_NOT_SET -> throw new IllegalArgumentException(
-                        "SystemUndelete without IdCase");
-            };
+                    .preHandle(setupRegularBuilder(context));
+            case TOKEN_PAUSE -> handlers.tokenPauseHandler()
+                    .preHandle(setupRegularBuilder(context));
+            case TOKEN_UNPAUSE -> handlers.tokenUnpauseHandler()
+                    .preHandle(setupRegularBuilder(context));
+
+            case UTIL_PRNG -> handlers.utilPrngHandler().preHandle(setupRegularBuilder(context));
+
+            case SYSTEMDELETE -> {
+                switch (context.getTxBody().getSystemDelete().getIdCase()) {
+                    case CONTRACTID -> handlers.contractSystemDeleteHandler()
+                            .preHandle(setupRegularBuilder(context));
+                    case FILEID -> handlers.fileSystemDeleteHandler()
+                            .preHandle(setupRegularBuilder(context));
+                    case ID_NOT_SET -> throw new IllegalArgumentException(
+                            "SystemDelete without IdCase");
+                }
+            }
+            case SYSTEMUNDELETE -> {
+                switch (context.getTxBody().getSystemUndelete().getIdCase()) {
+                    case CONTRACTID -> handlers.contractSystemUndeleteHandler()
+                            .preHandle(setupRegularBuilder(context));
+                    case FILEID -> handlers.fileSystemUndeleteHandler()
+                            .preHandle(setupRegularBuilder(context));
+                    case ID_NOT_SET -> throw new IllegalArgumentException(
+                            "SystemUndelete without IdCase");
+                }
+            }
 
             case NODE_STAKE_UPDATE, DATA_NOT_SET -> throw new UnsupportedOperationException(
                     "Not implemented");
+        }
+    }
+
+    private static SigTransactionMetadataBuilder setupRegularBuilder(
```

Comment:
Do we need to be that specific for the result type? Maybe `TransactionMetadataBuilder` is good enough?